File: android/engine/src/main/kotlin/ru/gigapisar/engine/Features.kt
package ru.gigapisar.engine

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.ln
import kotlin.math.log10
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow

/**
 * Звуковая волна → лог-мел-спектрограмма, ровно как ждёт энкодер GigaAM.
 *
 * Самое хрупкое место: разойтись здесь с оригиналом (server/giga_core.py,
 * web/giga/features.js) значит получить бессмыслицу. Поэтому всё шаг в шаг,
 * включая места, где числа округляются до float32 — numpy считает окно,
 * кадры, спектр и мел-полосы именно в такой точности.
 *
 * Преобразование Фурье — умножение на матрицу синусов и косинусов: длина
 * окна 320 не степень двойки, а на таких размерах матрица честна и быстра.
 */
data class FeatureConfig(
    val sampleRate: Int = 16000,
    val nMels: Int = 64,
    val nFFT: Int = 320,
    val winLength: Int = 320,
    val hopLength: Int = 160,
    val center: Boolean = false,
)

class Features(val cfg: FeatureConfig = FeatureConfig()) {
    val nFreqs = cfg.nFFT / 2 + 1
    private val window = FloatArray(cfg.winLength) { n -> (0.5 - 0.5 * cos(2.0 * PI * n / cfg.winLength)).toFloat() }
    private val cosT = DoubleArray(nFreqs * cfg.nFFT)
    private val sinT = DoubleArray(nFreqs * cfg.nFFT)
    private val fb: FloatArray

    init {
        val n = cfg.nFFT
        for (k in 0 until nFreqs) for (j in 0 until n) {
            val a = 2.0 * PI * ((k * j) % n) / n     // (k·n) mod N — угол маленький и точный
            cosT[k * n + j] = cos(a)
            sinT[k * n + j] = -kotlin.math.sin(a)
        }
        fb = melFilterbank(nFreqs, 0.0, cfg.sampleRate / 2.0, cfg.nMels, cfg.sampleRate)
    }

    /** Сколько кадров получится — та же формула, что в оригинале. */
    fun outLen(samples: Int): Int =
        if (cfg.center) samples / cfg.hopLength + 1 else (samples - cfg.winLength) / cfg.hopLength + 1

    class Result(val values: FloatArray, val frames: Int)

    /** Волна (16 кГц, float) → [nMels × кадры] подряд по строкам. */
    fun compute(wave: FloatArray): Result {
        val nFFT = cfg.nFFT; val hop = cfg.hopLength; val nMels = cfg.nMels
        var x = wave
        if (cfg.center) {
            val pad = nFFT / 2
            val padded = FloatArray(x.size + 2 * pad)
            for (i in 0 until pad) padded[i] = x[min(pad - i, x.size - 1)]
            System.arraycopy(x, 0, padded, pad, x.size)
            for (i in 0 until pad) padded[pad + x.size + i] = x[max(0, x.size - 2 - i)]
            x = padded
        }
        val n = max(0, (x.size - nFFT) / hop + 1)
        if (n == 0) return Result(FloatArray(0), 0)
        val frame = FloatArray(nFFT)
        val power = FloatArray(nFreqs)
        val out = FloatArray(nMels * n)
        for (f in 0 until n) {
            val off = f * hop
            for (j in 0 until nFFT) frame[j] = x[off + j] * window[j]
            for (k in 0 until nFreqs) {
                val row = k * nFFT
                var re = 0.0; var im = 0.0
                for (j in 0 until nFFT) {
                    val v = frame[j].toDouble()
                    re += v * cosT[row + j]
                    im += v * sinT[row + j]
                }
                power[k] = (re * re + im * im).toFloat()
            }
            for (m in 0 until nMels) {
                var s = 0.0
                for (k in 0 until nFreqs) s += power[k].toDouble() * fb[k * nMels + m].toDouble()
                val mel = min(max(s.toFloat(), 1e-9f), 1e9f)
                out[m * n + f] = ln(mel.toDouble()).toFloat()
            }
        }
        return Result(out, n)
    }

    companion object {
        private fun hzToMel(f: Double) = 2595.0 * log10(1.0 + f / 700.0)
        private fun melToHz(m: Double) = 700.0 * (10.0.pow(m / 2595.0) - 1.0)

        /** Треугольные фильтры, как torchaudio.functional.melscale_fbanks (norm=None). */
        fun melFilterbank(nFreqs: Int, fMin: Double, fMax: Double, nMels: Int, sampleRate: Int): FloatArray {
            val top = (sampleRate / 2).toDouble()
            val allFreqs = DoubleArray(nFreqs) { i -> top * i / (nFreqs - 1) }
            val mMin = hzToMel(fMin); val mMax = hzToMel(fMax)
            val fPts = DoubleArray(nMels + 2) { i -> melToHz(mMin + (mMax - mMin) * i / (nMels + 1)) }
            val fDiff = DoubleArray(nMels + 1) { i -> fPts[i + 1] - fPts[i] }
            val out = FloatArray(nFreqs * nMels)
            for (i in 0 until nFreqs) for (m in 0 until nMels) {
                val down = -(fPts[m] - allFreqs[i]) / fDiff[m]
                val up = (fPts[m + 2] - allFreqs[i]) / fDiff[m + 1]
                out[i * nMels + m] = max(0.0, min(down, up)).toFloat()
            }
            return out
        }
    }
}

File: windows/gigapisar/asr/features.go
// Звуковая волна → лог-мел-спектрограмма, шаг в шаг как server/giga_core.py,
// web/giga/features.js и android engine/Features.kt: та же точность float32
// в тех же местах, иначе на выходе не «чуть хуже», а бессмыслица.
package asr

import "math"

type FeatureConfig struct {
	SampleRate, NMels, NFFT, WinLength, HopLength int
	Center                                        bool
}

var DefaultFeatures = FeatureConfig{SampleRate: 16000, NMels: 64, NFFT: 320, WinLength: 320, HopLength: 160}

type Features struct {
	cfg    FeatureConfig
	nFreqs int
	window []float32
	cosT   []float64
	sinT   []float64
	fb     []float32
}

func NewFeatures(cfg FeatureConfig) *Features {
	f := &Features{cfg: cfg, nFreqs: cfg.NFFT/2 + 1}
	f.window = make([]float32, cfg.WinLength)
	for n := range f.window {
		f.window[n] = float32(0.5 - 0.5*math.Cos(2.0*math.Pi*float64(n)/float64(cfg.WinLength)))
	}
	N := cfg.NFFT
	f.cosT = make([]float64, f.nFreqs*N)
	f.sinT = make([]float64, f.nFreqs*N)
	for k := 0; k < f.nFreqs; k++ {
		for j := 0; j < N; j++ {
			a := 2.0 * math.Pi * float64((k*j)%N) / float64(N)
			f.cosT[k*N+j] = math.Cos(a)
			f.sinT[k*N+j] = -math.Sin(a)
		}
	}
	f.fb = melFilterbank(f.nFreqs, 0, float64(cfg.SampleRate)/2, cfg.NMels, cfg.SampleRate)
	return f
}

func hzToMel(f float64) float64 { return 2595.0 * math.Log10(1.0+f/700.0) }
func melToHz(m float64) float64 { return 700.0 * (math.Pow(10.0, m/2595.0) - 1.0) }

func melFilterbank(nFreqs int, fMin, fMax float64, nMels, sampleRate int) []float32 {
	top := float64(sampleRate / 2)
	all := make([]float64, nFreqs)
	for i := range all {
		all[i] = top * float64(i) / float64(nFreqs-1)
	}
	mMin, mMax := hzToMel(fMin), hzToMel(fMax)
	fPts := make([]float64, nMels+2)
	for i := range fPts {
		fPts[i] = melToHz(mMin + (mMax-mMin)*float64(i)/float64(nMels+1))
	}
	fDiff := make([]float64, nMels+1)
	for i := range fDiff {
		fDiff[i] = fPts[i+1] - fPts[i]
	}
	out := make([]float32, nFreqs*nMels)
	for i := 0; i < nFreqs; i++ {
		for m := 0; m < nMels; m++ {
			down := -(fPts[m] - all[i]) / fDiff[m]
			up := (fPts[m+2] - all[i]) / fDiff[m+1]
			v := math.Min(down, up)
			if v < 0 {
				v = 0
			}
			out[i*nMels+m] = float32(v)
		}
	}
	return out
}

// OutLen — сколько кадров энкодер ждёт для такого числа отсчётов.
func (f *Features) OutLen(samples int) int {
	if f.cfg.Center {
		return samples/f.cfg.HopLength + 1
	}
	return (samples-f.cfg.WinLength)/f.cfg.HopLength + 1
}

// Compute: волна → [nMels × кадры] подряд по строкам, число кадров.
func (f *Features) Compute(wave []float32) ([]float32, int) {
	nFFT, hop, nMels := f.cfg.NFFT, f.cfg.HopLength, f.cfg.NMels
	x := wave
	if f.cfg.Center {
		pad := nFFT / 2
		padded := make([]float32, len(x)+2*pad)
		for i := 0; i < pad; i++ {
			j := pad - i
			if j > len(x)-1 {
				j = len(x) - 1
			}
			padded[i] = x[j]
		}
		copy(padded[pad:], x)
		for i := 0; i < pad; i++ {
			j := len(x) - 2 - i
			if j < 0 {
				j = 0
			}
			padded[pad+len(x)+i] = x[j]
		}
		x = padded
	}
	n := (len(x)-nFFT)/hop + 1
	if len(x) < nFFT || n <= 0 {
		return nil, 0
	}
	frame := make([]float32, nFFT)
	power := make([]float32, f.nFreqs)
	out := make([]float32, nMels*n)
	for fr := 0; fr < n; fr++ {
		off := fr * hop
		for j := 0; j < nFFT; j++ {
			frame[j] = x[off+j] * f.window[j]
		}
		for k := 0; k < f.nFreqs; k++ {
			row := k * nFFT
			var re, im float64
			for j := 0; j < nFFT; j++ {
				v := float64(frame[j])
				re += v * f.cosT[row+j]
				im += v * f.sinT[row+j]
			}
			power[k] = float32(re*re + im*im)
		}
		for m := 0; m < nMels; m++ {
			var s float64
			for k := 0; k < f.nFreqs; k++ {
				s += float64(power[k]) * float64(f.fb[k*nMels+m])
			}
			mel := float32(s)
			if mel < 1e-9 {
				mel = 1e-9
			}
			if mel > 1e9 {
				mel = 1e9
			}
			out[m*n+fr] = float32(math.Log(float64(mel)))
		}
	}
	return out, n
}
